auth/certificate_authenticator.cc

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "auth/certificate_authenticator.hh"
+
+#include <regex>
+
+#include "utils/class_registrator.hh"
+#include "replica/database.hh"
+#include "cql3/query_processor.hh"
+#include "db/config.hh"
+
+static const auto CERT_AUTH_NAME = "com.scylladb.auth.CertificateAuthenticator";
+const std::string_view auth::certificate_authenticator_name(CERT_AUTH_NAME);
+const std::string subject_key = "subject";
+
+static logging::logger clogger("certificate_authenticator");
+
+static const class_registrator<auth::authenticator
+    , auth::certificate_authenticator
+    , cql3::query_processor&
+    , ::service::migration_manager&> cert_auth_reg(CERT_AUTH_NAME);
+
+auth::certificate_authenticator::certificate_authenticator(cql3::query_processor& qp, ::service::migration_manager&)
+    : _subject_expr([&] {
+        auto& conf = qp.db().get_config();
+        auto query = conf.auth_certificate_role_query();
+
+        if (query.empty()) {
+            throw std::invalid_argument("No role extraction query specified.");
+        }
+
+        try {
+            std::regex ex(query);
+            if (ex.mark_count() != 1) {
+                throw std::invalid_argument("Role query must have exactly one mark expression");
```

Comment:
If a user wants to:
- authenticate admins via their emails
- authenticate servers via their CN/SAN

then it looks like what really matters is that there is only one group that matches when resolving the regex (actually even then we could probably take the first one), but there may be several of them defined, e.g. https://regex101.com/r/Slki3N/3

That doesn't seem needed anymore if we make the `auth_certificate_role_query` a list.

Reply:
The check should be `ex.mark_count() == 0`, and we should allow for arbitrary match clauses, picking the first one that fits.